Update the now playing UI

saecki opened the pull request:
This updates the now playing fragment, adds a details dialog and includes some other changes:

- converted the `PlayerFragment` to Kotlin
- added `duration` and `year` properties to `CollectionItem`
- refactored out `BackdropMenu`, all logic now sits inside `BackdropLayout`

The details dialog could easily be reused in the collection or queue to display more information about songs or directories.

Dark | Light
--- | ---
![image](https://user-images.githubusercontent.com/43008152/116466690-1d385780-a86f-11eb-86c6-0e311bf175ed.png) | ![image](https://user-images.githubusercontent.com/43008152/116466761-2fb29100-a86f-11eb-8c40-00d8491a3aec.png)
![image](https://user-images.githubusercontent.com/43008152/116466866-4f49b980-a86f-11eb-8486-eb668b37a687.png) | ![image](https://user-images.githubusercontent.com/43008152/116468317-ff6bf200-a870-11eb-98f9-198363cbf568.png)

saecki committed:
redesing the player fragment

saecki committed:
move playback buttons below seekbar

saecki committed:
fix progressbar mode and seekbar NaN bug

saecki committed:
convert player fragment to kotlin

saecki committed:
add details dialog to the player fragment

saecki committed:
add duration and year properties

saecki committed:
add scroll view to the player details dialog

saecki committed:
improve resizing of player fragment and refactor out backdrop menu

saecki committed:
fix artwork alignment in player fragment

saecki committed:
create top level show details dialog function

saecki committed:
fix player fragment controls min height

saecki committed:
clean up PlayerFragment

saecki committed:
prevent accidentally clicking backdrop menu through front layout

saecki committed:
make showDetails an extension function of context

saecki committed:
use details prefix for details strings

saecki committed:
update dependencies

saecki committed:
use lazy initialized unknown string in details and player fragment

agersant commented on app/src/main/java/agersant/polaris/api/remote/PolarisExoPlayerDataSourceFactory.java:
```diff
@@ -66,7 +66,7 @@ public void onTransferEnd(DataSource source, DataSpec dataSpec, boolean isNetwor
 
         PolarisExoPlayerHttpDataSource(OfflineCache offlineCache, RequestProperties requestProperties, PolarisExoPlayerTransferListener listener, File scratchLocation, CollectionItem item) {
-            super("Polaris Android", null, DEFAULT_CONNECT_TIMEOUT_MILLIS, DEFAULT_READ_TIMEOUT_MILLIS, true, requestProperties);
+            super("Polaris Android", DEFAULT_CONNECT_TIMEOUT_MILLIS, DEFAULT_READ_TIMEOUT_MILLIS, true, requestProperties); //TODO: use factory instead of deprecated constructor
```
Assuming it isn't a very large change, could you do this upgrade so we don't keep deprecated code around?

Reading ExoPlayer release notes, I see they now have a playlist API which might be good to investigate. I think Polaris could benefit from it in terms of achieving gapless playback and potentially having less Polaris code. I'll write a new issue for that one, no need to worry about it in this PR.

saecki replied:
I have rough plans on upgrading to `androidx.media2` which if I remember correctly `Exoplayer` has an extension for, that integrates quite nicely. The playlist API sounds interesting and I will definitely take a look at that.
The problem is that the deprecated constructor is being replaced by a Factory. Since were extending `DefaultHttpDataSource` that seemed like too much trouble right now.

agersant commented on app/src/main/res/layout/view_song_details.xml:
```diff
@@ -0,0 +1,154 @@
+                style="@style/DetailsSpace"
+                android:layout_width="match_parent" />
+
+            <TextView
```
I would suggest re-ordering these as follows. It's less hierarchical but it seems more intuitive to me in terms of relative importance:

- Title
- Album
- Year
- Artist
- Album Artist 
- Track Number
- Total Tracks
- Duration
- Path

Displaying the path also seems a bit ugly, but I'm OK to keep it if you have a use case.

saecki replied:
Done

agersant commented:
I'm very happy to get more info on the player page. Especially duration and current time have been cruelly missing for too long!

I do have some reserves on the visuals, I find the `Information` icon to be a bit too bold and breaking the balance of the screen. I would suggest using the top right corner for it. I also tried below the album cover (right-aligned) but it didn't feel as good to me.

Another change I did is adjust the spacing to give the info more breathing room:

![Untitled](https://user-images.githubusercontent.com/817256/116794966-30714000-aa86-11eb-96db-611bfbc827e3.png)

saecki committed:
change order of elements in details dialog

saecki committed:
now playing ui changes

- move show details icon to the toolbar
- use rounded icons for playback controls
- add more spacing between artist and seekbar
- remove path from details dialog

saecki commented:
> I find the Information icon to be a bit too bold and breaking the balance of the screen. I would suggest using the top right corner for it
- I moved the details icon to the toolbar menu and used the filled instead of the outlined one.

> Another change I did is adjust the spacing to give the info more breathing room  

- Done

- I also changed the playback control icons to rounded ones. Let me know if you prefer these or the sharp ones.
- After looking at it again I don't know why I chose to display the path in the details dialog in the first place and removed it.

saecki commented:
Light | Dark
--- | ---
![image](https://user-images.githubusercontent.com/43008152/117063970-d8f0ff80-ad25-11eb-911f-9ba64b62a79c.png) | ![image](https://user-images.githubusercontent.com/43008152/117064000-e27a6780-ad25-11eb-9b4e-b9a333ef9655.png)
![image](https://user-images.githubusercontent.com/43008152/117064190-2a00f380-ad26-11eb-9678-e321e828e7fa.png) | ![image](https://user-images.githubusercontent.com/43008152/117064227-3422f200-ad26-11eb-8a92-a567facca176.png)

saecki committed:
fix warning and clean up MainActivity

saecki committed:
remove unused SquareLayout

saecki committed:
refactor BackdropLayout

saecki committed:
handle player options item selected better

saecki committed:
downgrade exoplayer core and upgrade flac extension

saecki commented:
To fix flac playback I had to downgrade exoplayer to version `2.11.8` and used the latest autobuild flac extension version `2.11.4`.

saecki committed:
update dependencies

- update kotlin to 1.5.0
- update android gradle plugin to 4.2.0
- update ndk version to 22
- update exoplayer and flac extension to 2.13.3
- replace jcenter with mavenCentral

saecki committed:
fix warnings

saecki commented:
I forked and updated PaulWoitaschek's ExoPlayer-Extensions repo for the flac extension to `2.13.3` and added that as a dependency for now. There is an open PR though, as soon as that is resolved we can go back to using his one,

saecki committed:
stop seekbar updates when player fragment is not visible

agersant commented:
> I also changed the playback control icons to rounded ones.

Looks great!

> I forked and updated PaulWoitaschek's ExoPlayer-Extensions repo for the flac extension to 2.13.3 and added that as a dependency for now. There is an open PR though, as soon as that is resolved we can go back to using his one.

May be worth opening a ticket on ourselves for tracking.

I loaded the changes on my device, will spend some time listening to music this weekend. A couple things that I did spot already:

- Loading indicator on the left is a great addition
- When skipping to a different song, the song duration briefly displays `0:00`. May look better if it was empty instead?
- Seeing `unknown` in the details popup can look a bit sad. For example, I intentionally don't populate Album Artist for records with a single artist, or disc number for records with a single disc in my collection. This is making me think maybe we should simply omit blank fields?

saecki committed:
update deprecated exoplayer code

saecki committed:
hide duration when loading new song

saecki commented:
> Seeing unknown in the details popup can look a bit sad. For example, I intentionally don't populate Album Artist for records with a single artist, or disc number for records with a single disc in my collection. This is making me think maybe we should simply omit blank fields?

I actually like having a consistent amount of metadata fields, helps me to find what I'm looking for more quickly. The `[unknown]` is quite prominent though, maybe something more subtle will do. I thought of some icon or unicode character but couldn't really find something fitting. How about just `unknown` and graying out the text?.

Dark | Light
--- | ---
![image](https://user-images.githubusercontent.com/43008152/118402568-c56d5f00-b66a-11eb-9cb3-37732498ec4c.png) | ![image](https://user-images.githubusercontent.com/43008152/118402602-e1710080-b66a-11eb-80e9-b6f7af3269f5.png)

saecki committed:
make unknown detail values more subtle

saecki committed:
omit non existent details properties

saecki commented:
In regards to the additional properties, that are being added to polaris in the moment, I might also prefer omitting fields that are not present. So here we go.

agersant commented:
Updated code look all good. My testing hasn't revealed any additional issues, will do one more round of testing on device with latest and then we're good to merge. 🥳

agersant commented:
Ship it!

I did find one bug in my testing but I don't think it's related to this PR, I'll open a separate issue.
